api/custom_index.py

```diff
@@ -0,0 +1,215 @@
+"""A custom index view customizes a library's 'front page' to serve
+something other than the default.
+"""
+
+from nose.tools import set_trace
+
+from flask import Response
+from flask_babel import lazy_gettext as _
+
+from sqlalchemy.orm.session import Session
+
+from config import CannotLoadConfiguration
+from core.model import (
+    get_one,
+)
+from core.lane import Lane
+from core.model import (
+    ConfigurationSetting,
+    ExternalIntegration,
+)
+from core.util.opds_writer import OPDSFeed
+
+class CustomIndexView(object):
+    """A custom view that replaces the default OPDS view for a
+    library.
+
+    Any subclass of this class must define PROTOCOL and must be
+    passed into a CustomIndexView.register() call after the class
+    definition is complete.
+
+    Subclasses of this class are loaded into the CirculationManager, so they
+    should not store any objects obtained from the database without
+    disconnecting them from their session.
+    """
+    BY_PROTOCOL = {}
+
+    GOAL = "custom_index"
+
+    @classmethod
+    def register(self, view_class):
+        protocol = view_class.PROTOCOL
+        if protocol in self.BY_PROTOCOL:
+            raise ValueError("Duplicate index view for protocol: %s" % protocol)
+        self.BY_PROTOCOL[protocol] = view_class
+
+    @classmethod
+    def unregister(self, view_class):
+        """Remove a CustomIndexView from consideration.
+        Only used in tests.
+        """
+        del self.BY_PROTOCOL[view_class.PROTOCOL]
+
+    @classmethod
+    def for_library(cls, library):
+        """Find the appropriate CustomIndexView for the given library."""
+        _db = Session.object_session(library)
+        integration = ExternalIntegration.one_for_library_and_goal(
+            _db, library, cls.GOAL
+        )
```

Comment:
This is not really an "external" integration since the custom view code is in the circ manager, but I guess the same is true for something like the simple authentication provider.